Add cost inference of fwd sparse operators and sparse adagrad

jspark1105 opened the pull request:
Summary: We should also add cost inference for sparse operators in backward pass later.

Differential Revision: D8789240

jspark1105 commented:
@salexspb @mdschatz

salexspb commented on caffe2/operators/segment_reduction_op.cc:
```diff
@@ -2,6 +2,47 @@
+  if (N == 0) {
+    return c;
+  }
+  uint64_t D = nElemFromDim(data) / N;
```
why not do from dim 1 and then not divide by N ? The function name is misleading, it should really accept dim as the second argument

jspark1105 replied:
I wasn't able to find a function that computes size from a certain dimension. You're right we should change nElemFromDim function. Can we do that in a separate PR? :)

salexspb commented on caffe2/operators/segment_reduction_op.cc:
```diff
@@ -2,6 +2,47 @@
+  c.bytes_read = indices_size *
+          (D * sizeof(data.data_type()) + sizeof(indices.data_type())) +
+      M * sizeof(lengths.data_type());
+  c.params_bytes = N * D * sizeof(data.data_type());
```
I think there is a mess here with param_bytes. Is it expected to be how much we read parameters or their total size? 

I think we want to have total params read, activations bytes read and activation bytes written
cc @highker

jspark1105 replied:
params_byte is for total parameter size. Yes, it would be useful to separate params read, activation bytes read, and activation bytes written, but probably in a separate PR :)

jspark1105 replied:
Sounds good.

jspark1105 replied:
Do we want to make this change of params_bytes including updating the comment in this PR or as a separate?

salexspb replied:
sure, lets update the comment :) cc @highker

highker replied:
This part looks quite reasonable to me :D

mdschatz commented on caffe2/core/operator_schema.h:
```diff
@@ -478,9 +478,9 @@ inline vector<TIndex> GetDimsVector(const TensorShape& shape) {
-  for (int i = 0; i < X.dims_size(); ++i) {
+inline uint64_t nElemFromDim(const TensorShape& X, int dim = 0) {
+  uint64_t nElem = X.dims_size() > dim ? 1 : 0;
+  for (int i = dim; i < X.dims_size(); ++i) {
```
Could we use `size_from_dim_` here?

jspark1105 replied:
One issue is TensorShape is not a vector and operator_schema.h doesn't include tensor.h as far as I can tell.

salexspb replied:
btw, 0! == 1. Don't we want to report 1 for overshooting dims? Or may be even throw in this case? 0 is kinda surprising.

jspark1105 replied:
Since this is a utility function we may want to be conservative about throwing an error? For example, what if by some reason one uses an empty FC/Conv bias as an empty tensor? So maybe a warning?

salexspb replied:
okay, size_from_dim in tensor.h returns 1 for such case. We probably should be consistent. This is also consistent with 0!

pietern commented:
@pytorchbot retest this please

jspark1105 commented:
@pytorchbot retest this please

jspark1105 commented:
@salexspb  and @mdschatz , I hope addressed all the comments. If so, can we land?

mdschatz commented:
I'm fine with these changes.  Happy to land.

jspark1105 commented:
Cool. Can anyone accept the diff? @mdschatz @salexspb

bddppq commented:
@pytorchbot retest this please

jspark1105 commented:
@highker

highker commented on caffe2/core/operator_schema.h:
```diff
@@ -478,9 +478,9 @@ inline vector<TIndex> GetDimsVector(const TensorShape& shape) {
-  uint64_t nElem = X.dims_size() > 0 ? 1 : 0;
-  for (int i = 0; i < X.dims_size(); ++i) {
+inline uint64_t nElemFromDim(const TensorShape& X, int dim = 0) {
+  uint64_t nElem = 1;
```
Are we dropping the `nElem == 0` case?

jspark1105 replied:
Yes, to simplify code based on discussion with Alexander.

highker commented on caffe2/sgd/adagrad_op.cc:
```diff
@@ -37,6 +37,30 @@ Optionally returns effective_lr and update as well.
+  uint64_t grad_size = nElemFromDim(grad);
+
+  OpSchema::Cost c;
+  c.flops = grad_size * 7; // optimistically count sqrt as one flop
```
hmm... why is `7`?

jspark1105 replied:
float hi = nh[i] = h[i] + gi * gi;
    nw[i] = w[i] + lr * gi / (std::sqrt(hi) + epsilon);

Here, we're doing 2 multiplications, 3 additions, 1 division, and 1 sqrt.

highker replied:
I see, could you please add this to the comment?

highker commented on caffe2/operators/segment_reduction_op.cc:
```diff
@@ -2,6 +2,47 @@
+          (D * sizeof(data.data_type()) + sizeof(indices.data_type())) +
+      M * sizeof(lengths.data_type());
+  c.params_bytes = N * D * sizeof(data.data_type());
+  if (use_weight) {
```
hmmm... I'm not quite familiar with this but isn't `weight` a param?

jspark1105 replied:
This class implements 2 op types first without weight input (SparseLengthsSum) and second with weight (SparseLengthsWeightedSum). This if statement is for SparseLengthsWeightedSum.

jspark1105 commented:
@pytorchbot retest this please

jspark1105 commented:
@highker please merge this PR. I don't have a write access :)

highker commented:
@jspark1105 me neither.... @orionr could you help? Thanks

jspark1105 commented:
@orionr @bddppq @Yangqing can anyone help merging this?

orionr commented:
@jspark1105 and @highker if you don't have the "Import to Phabricator" button here that means you need to install FaceHub. Do "s facehub" internally and you should find steps. Let me know if you have issues. In the meantime, I've created the diff and we should be able to land it through fbcode and it will then arrive in master.

jspark1105 commented:
@pytorchbot retest this please

jspark1105 commented:
@pytorchbot retest this please

jspark1105 commented:
@pytorchbot retest this please

jspark1105 committed:
Add cost inference of fwd sparse operators and sparse adagrad (#9314)

Summary:
We should also add cost inference for sparse operators in backward pass later.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/9314

Differential Revision: D8789240

fbshipit-source-id: b5b32ab52741b3fb81f6af96777703bd59b24c37
